Color support in 3D rendering (includes OFF color import / export, green subtracted faces)

ochafik opened the pull request:
This is an improved / cleaned up subset of https://github.com/openscad/openscad/pull/5180 to support colors in 3D models (https://github.com/openscad/openscad/issues/5065 and #5047).

- OFF format import / export only for now (will send more - incl. X3D & glTF - through Assimp in subsequent PR)
- Enabled by feature `render-colors`, and needs `manifold` to preserve colors across CSG ops. Works w/ `lazy-union`.
- Uses Manifold's OriginalID mechanism to trace original colors as described by @elalish in https://github.com/openscad/openscad/issues/5065#issuecomment-2168257810
- Adds optional per-face indexed colors to PolySet. Didn't adopt the runIndex mechanism from manifold::MeshGL as it's a bit overkill for now, but we may want to look into it to speed up rendering later.
- For multimaterial 3d printing, solid colors might have to be implemented separately ~~(either in OpenSCAD using CGAL's tetrahedral mesh creation routines, or~~ in the slicers themselves, cf. https://github.com/assimp/assimp/issues/5629).

Test it on the Web Playground: http://ochafik.com/openscad2/

TODO:
- [x] add test cases (w/ pngs & `.off` contents)
- [x] Render uncolored faces subtracted faces as green w/ Manifold to match preview
- [ ] add examples

Planned follow ups (from @kintel 's suggestions below):
- Add small indicator in view to distinguish renders from previews
- Scale up to large number of colors

kintel commented:
What do you think about the idea of using this feature to make rendering with default colors use a different color for negative object (e.g. green subtracted from yellow), to make Manifold rendering look the same as CGAL rendering?

This may only have value for rendering purposes though.

kintel commented:
Now that preview and render looks the same, we may need to add a small indicator in the view, so people know what they're looking at and don't think they misclicked.

kintel commented:
I see you added 3MF export. Are you planning on doing import as well in this PR? If not, it could be worth postponing that to the next CL, as testing becomes significantly easier if both import and export is implemented.

kintel commented:
In terms of tests, we can probably utilize some existing tests with colors.
A quick test uncovered that this renders incorrectly (yellow color disappears in preview mode):
```
difference() {
    color("yellow") render() difference() {
      color("red") sphere(10);
      color("blue") cube(10);
    }
    color("green") cylinder(r=3, h=20, center=true);
}
```

kintel commented on src/geometry/PolySet.cc:
```diff
@@ -106,6 +106,13 @@ void PolySet::transform(const Transform3d& mat)
+void PolySet::setColor(const Color4f& c) {
+  colors.clear();
+  colors.push_back(c);
+  color_indices.clear();
+  color_indices.resize(indices.size(), 0);
```
`color_indices.assign(indices.size(), 0)`

kintel commented on src/geometry/PolySet.cc:
```diff
@@ -106,6 +106,13 @@ void PolySet::transform(const Transform3d& mat)
 }
 
+void PolySet::setColor(const Color4f& c) {
+  colors.clear();
+  colors.push_back(c);
```
`colors.assign(1, c)` or `colors = {c}`

kintel reviewed (commented):
Just walking the boundary and giving it a quick spin. Will continue my review when I find more time :)

kintel commented on src/geometry/manifold/ManifoldGeometry.h:
```diff
@@ -16,7 +19,7 @@ class ManifoldGeometry : public Geometry
 
   ManifoldGeometry();
-  ManifoldGeometry(const std::shared_ptr<const manifold::Manifold>& object);
+  ManifoldGeometry(const std::shared_ptr<const manifold::Manifold>& object, const std::map<uint32_t, Color4f> & originalIDToColor = std::map<uint32_t, Color4f>());
```
Nit: you could probably just do `= {}` instead of assigning a fully specified type

ochafik commented:
Thanks @kintel !!

> What do you think about the idea of using this feature to make rendering with default colors use a different color for negative object (e.g. green subtracted from yellow), to make Manifold rendering look the same as CGAL rendering?
> This may only have value for rendering purposes though. One benefit could be that we'd get a lot of tests for free.

Oh that's a cool idea! Added this and your other suggestions to a list of follow ups, I'd probably bundle this one w/ rendering support of background / debug modifiers in a $preview mode to bring more proper preview to the web playground.

> I see you added 3MF export. Are you planning on doing import as well in this PR? If not, it could be worth postponing that to the next PR, as testing becomes significantly easier if both import and export is implemented.

Ah good point, removed / [parked here](https://github.com/openscad/openscad/commit/fc61ba6f1337cdfe4b265367fd305e29a4beb1a9).

> In terms of tests, we can probably utilize some existing tests with colors. A quick test uncovered that this renders incorrectly (yellow color disappears in preview mode):
> 
> ```
> difference() {
>     color("yellow") render() difference() {
>       color("red") sphere(10);
>       color("blue") cube(10);
>     }
>     color("green") cylinder(r=3, h=20, center=true);
> }
> ```

Ah cool! I've necromanced a fix for that (shuffled GeometryEvaluator::evaluateGeometry), but can't get a proper preview to png to test it (this example exported to png has a solid green cylinder). Tried `--preview=throwntogether`, any idea?

kintel commented:
> Ah cool! I've necromanced a fix for that (shuffled GeometryEvaluator::evaluateGeometry), but can't get a proper preview to png to test it (this example exported to png has a solid green cylinder). Tried `--preview=throwntogether`, any idea?

Note sure what you mean.. Preview to PNG would be `openscad test.scad -o out.png --enable=manifold  --enable=render-colors`

kintel commented:
..but Render to PNG (as opposed to preview) would be `openscad test.scad -o out.png --enable=manifold --enable=render-colors --render`

kintel commented on tests/CMakeLists.txt:
```diff
@@ -1044,8 +1054,9 @@ add_cmdline_test(manifold-stlexport    EXPERIMENTAL OPENSCAD SUFFIX stl FILES ${
-add_cmdline_test(color-offimport       EXPERIMENTAL OPENSCAD SUFFIX png FILES ${EXPORT_OFFIMPORT_TEST_FILES} ARGS --enable=colors)
+add_cmdline_test(color-offexport       EXPERIMENTAL OPENSCAD SUFFIX off FILES ${EXPORT_OFFEXPORT_TEST_FILES} ARGS --enable=manifold --enable=render-colors --enable=predictible-output --render)
+add_cmdline_test(color-offimport       EXPERIMENTAL OPENSCAD SUFFIX png FILES ${EXPORT_OFFIMPORT_TEST_FILES} ARGS --enable=render-colors)
+add_cmdline_test(colorpngtest          EXPERIMENTAL OPENSCAD SUFFIX png FILES ${COLOR_TEST_FILES} ARGS --enable=render-colors)
```
I would add `--render` here, otherwise you'll be testing previews.

ochafik replied:
ah!!

kintel commented on tests/CMakeLists.txt:
```diff
@@ -1035,6 +1053,10 @@ add_cmdline_test(stlexport             EXPERIMENTAL OPENSCAD SUFFIX stl FILES ${
 add_cmdline_test(manifold-stlexport    EXPERIMENTAL OPENSCAD SUFFIX stl FILES ${EXPORT_STL_TEST_FILES} EXPECTEDDIR stlexport ARGS --enable=predictible-output --enable=manifold --render)
 add_cmdline_test(objexport             EXPERIMENTAL OPENSCAD SUFFIX obj FILES ${EXPORT_OBJ_TEST_FILES} ARGS --render --enable=predictible-output)
 add_cmdline_test(3mfexport             EXPERIMENTAL OPENSCAD SUFFIX 3mf FILES ${EXPORT_3MF_TEST_FILES} ARGS --enable=predictible-output)
+add_cmdline_test(offexport             EXPERIMENTAL OPENSCAD SUFFIX off FILES ${EXPORT_OFFEXPORT_TEST_FILES} ARGS --enable=predictible-output --render)
+add_cmdline_test(color-offexport       EXPERIMENTAL OPENSCAD SUFFIX off FILES ${EXPORT_OFFEXPORT_TEST_FILES} ARGS --enable=manifold --enable=render-colors --enable=predictible-output --render)
+add_cmdline_test(color-offimport       EXPERIMENTAL OPENSCAD SUFFIX png FILES ${EXPORT_OFFIMPORT_TEST_FILES} ARGS --enable=render-colors)
```
This doesn't looks right: The test names indicate OFF export/import, but you're just doing a straight-up OpenSCAD render to PNG. To test an export/import round-trip, use `EX_IM_PNGTEST_PY`

ochafik replied:
shuffled the tests a bit, hope it makes more sense now

ochafik commented:
> > Ah cool! I've necromanced a fix for that (shuffled GeometryEvaluator::evaluateGeometry), but can't get a proper preview to png to test it (this example exported to png has a solid green cylinder). Tried `--preview=throwntogether`, any idea?
> 
> Note sure what you mean.. Preview to PNG would be `openscad test.scad -o out.png --enable=manifold --enable=render-colors`

@kintel I was confused about the `--preview` flag. Have now added the test, normal preview is fine, it's just the `--preview=throwntogether` that [looked funny](https://github.com/openscad/openscad/pull/5185/files#diff-7fd4542c55e81a1b2adb5fabc8c044af621b2ecfae91fd1584efe9b04a050cd8).

kintel commented:
> @kintel I was confused about the `--preview` flag. Have now added the test, normal preview is fine, it's just the `--preview=throwntogether` that [looked funny](https://github.com/openscad/openscad/pull/5185/files#diff-7fd4542c55e81a1b2adb5fabc8c044af621b2ecfae91fd1584efe9b04a050cd8).

The purpose of `--preview=throwntogether` is the same as F12 Throwntogether in the GUI; it renders all objects (both positive and negative) as positive, as a debugging tool for cases where preview and render gives bad/slow/confusing results. These days it has limited usage; the only strong remaining use-case is that we render all polygons with two-sided material in that mode, making it easier to spot polygon winding issues.

ochafik commented:
Tests are now *mostly green* (only failure seems unrelated)

ochafik commented:
> Tests are now _mostly green_ (only failure seems unrelated)

I've been greedy and somehow prioritized green faces over green tests. Hope to stabilize the CI again soon.

kintel commented:
This is all fantastic stuff!
I'm not even sure we need a separate experimental feature for it, as we can align this with the manifold feature itself. ..but let's focus on getting it merged first - clean up can happen in the green.

kintel commented on src/geometry/manifold/ManifoldGeometry.cc:
```diff
@@ -106,11 +113,80 @@ std::shared_ptr<const PolySet> ManifoldGeometry::toPolySet() const {
+      const auto id = mesh.runOriginalID[run];
+      const auto end = mesh.runIndex[run + 1];
+      const size_t numTri = (end - start) / 3;
+      assert(numTri > 0);
```
Reading Manifold source code, It looks like it's perfectly valid that `runIndex` entries have zero triangles, in cases where an object didn't contribute to the final mesh.

kintel replied:
I do wonder, though, why you're not seeing this assertion on your end?

elalish replied:
Oh right, I think I use that to ensure I get the combined transform for `show()` and `only()` akin to `#` and `*`.

ochafik replied:
Couldn't reproduce this on Mac yet (w/ CGAL 5.4.5; I suppose different versions could matter here since it's about minkowski), my code in createManifoldFromTriangularPolySet might be dodgy.

kintel replied:
My test was on macOS w/CGAL 5.5 (built from source using scripts/macosx-build-dependencies.sh).

kintel replied:
Can also reproduce without using minkowski

kintel replied:
Just as a test, I removed the assert, and the above example produces some interesting artifacts:

<img width="809" alt="Screenshot 2024-07-01 at 21 11 46" src="https://github.com/openscad/openscad/assets/186094/5a2a9c24-40ff-468d-9c8e-2b213f6f742e">

ochafik replied:
Is this on a (k)Intel Mac? Did you make sure the manifold submodule is updated?

kintel replied:
M1 mac - I'll double-check by updating and building from scratch

kintel replied:
Build from scratch; still same issue (assert and triangle coloring artifact)

kintel replied:
This example also behaves oddly on the web playground

kintel replied:
Some digging I realize my example was bogus (ambigous source triangles), so pls. disregard the screenshot above. However, the z tearing I see in the web playground it worth looking into.

kintel replied:
I managed to reproduce the assert failure on a fresh VM running macOS 14.4 and the latest Xcode.
I think the main difference from your setup may be that I've build the dependencies using scripts/macosx-build-dependencies.sh on both machines. I wonder if there is some minor difference in what libs get pulled in. ..or a bug surfacing in a 3rd party lib.

kintel replied:
I can also reproduce the assertion failure when using homebrew dependencies.
Not sure what next to try, perhaps you're on an older compiler? I'm on `Apple clang version 15.0.0 (clang-1500.1.0.2.5)`

kintel replied:
The web playground z buffer tearing feels like lazy union was used to produce two partially overlapping objects. If that's the case, it's WAI.

ochafik replied:
Ah of course. Lazy union can’t currently be disabled, I need to expose the options somewhere.

kintel replied:
Yeah, so the remaining puzzling issue here is why the assertion failure is not consistent across environments.

ochafik commented:
Quick note that this can be now tested on the main Web Playground which also just got a long list of other upgrades (customizer, exports to multiple formats, modelviewer, tweaks to 3mf output to make it printable by multimaterial printers, etc; to be upstreamed soon) 

http://ochafik.com/openscad2/

kintel commented on src/geometry/manifold/ManifoldGeometry.cc:
```diff
@@ -168,7 +168,9 @@ std::shared_ptr<PolySet> ManifoldGeometry::toPolySet() const {
       const auto end = mesh.runIndex[run + 1];
       const size_t numTri = (end - start) / 3;
-      assert(numTri > 0);
+      if (numTri == 0) {
```
Did you manage to reproduce this? I think this is the correct fix, but it's still a bit puzzling why it didn't break anywhere except on my machine (unless everyone else builds without assertions..)

ochafik replied:
I just reproduced it w/ a different model (while testing [rendering of modifiers](https://ochafik.com/openscad2/#H4sIAAAAAAAAE+V97ZLbOJLgq2RrZ2bFKuqDkqrclpv2lsvuHu/4K2z3zM5K2hqKhCTYFMghqaqSux1x/y7uJS76Fe7v/bl36Se5yEwABPXhLs/2zF3EOma6SDCRSCQSiUQikfqhlUdFtC5b4x9aUVzJa/E6qlatcauXp9F2WWQblXTLOEpafmshompTiLI1nrTWkZKLLKXiqKw6cbls+a00+rjtbJTMVMtvRWUp13nLbxVCJaLoxFmaFWX9vs4SuZCiKFszv3UdFUTDbxaqNT4f+a0/RoWM5qlojatiI/zW21yqeXbbGgd+6+VmPcd6+Pw8motUJFfXUbpB0oK+30Icg77f+qOIq6wYtcaTYOAPR/7o3B+MZqY8oHL7OjBgtmToVJz5rUTMN8vWeBGlpfjkt8psU8TEjB9a+VGexZmqhKpa49ZChYP+A+j1YDIcj/qzqZqqRC4WohAqFm0PfpgqAIDfxpu5aAdnPsRCVaIIsf/eA/5Y5itRiHbQ9+E3CxUuFH741Prk1zRkuVDYeE/cRus8FWXvcVTKuOxdvv1un6q/nYK7tHyRXEcqFknvu6ePD3LkzDBkyAzpncDkSZHlkGQ3CubZ7XgGJ72p6vUgztbzDItgkRWgSASmSotCOGBEfR8CHwY+DBndbrWyKqRaTtVb+luG09Yiy6YtrrzIMh/mUYH/+ajrpyxeB1svp6opfWGgBzjoj5/7MOiPX/gw7I//7flnsRmiGsjCaeutIezt+O06SlMfXoxfiERu1j48Hz+PiqXQTJvA21QmooCZZlbJrzcyWYqqwTD9JRyOLKlBXzO/V1YiN3UblSqRcwvM6El/fFZXO4HJ5UrEH+bZLY8WokpEGRcyr2SmpspMZggBZemBplrP6boSlFwCN7JaAVMjPwoIpkrDQghnD8A0+07cVk0EWHJktMNpayXSFIfb1H+bi1hGKVzTfDei1sDBn5igVJQlVKtIQVaA+OsmSqHKYAQiFWuhqnKqtGYJUbFoRg3GZ8gnrWRCrWMOfRyGtbo59H0UNvRYE4Q79HuZJEJxR0hfvciSBtf19DXK486aA7X1JhXl/izu9WD3O3SAasKmjJYCsgXorz7IhQ+0EKACK3u/WUR64C5XkVoKqFayRKaSlsWHQqyzaywXsBJpLgpYimwtqmKre9iQqR58l2bzKIVClFm6Yen7zaKEEPrd4AGgyD/J1D9XsBRKFFEloMSphfIexaLSw9vvBrBeY82IBQ4O1UxxBhYQqWUqdMUzSMSyEKLU1LyIpHIJPqBtJarYUsRIa12K/+ZZsm0bvdsAzQpb/on/rLJUlFT4STf9e+JWOVVyAW3ilKc5yHDM9awQhj4ZQ17ItUQjoOzybFqJUkAcKZgL0Ot1gqMSFwI5cI3zGgkvcZBxjBZSRSlk8/cirrANHnjdFdM5EoD2tPU43Yhpy9tZU5xqjQ7v1n4jEqx8bLFy8SB3XhFN+2ieyzUREW9TiXZJexUO+j4U4RGUvR6gOss2pZFq1hOy3ESp/CjVkslei0Qij+JsnWeK1cN+x0Tya0rCbp8v2h4UWRVVoj3p+/f7fn/mNZjxoAH+2AFH4F8Av2zvfW18L+tOkE3Y7JOm70Gz5PFeyeXRDhphNlh7PXimlCj0uEApVSxAVv9cQqbSLSghEpGAVKVMbHWuqmukUuHMrOUzij80ZKMIAx9WIdpfO7KhraM4SkW73z1rdLUqIlWmxNbOEPnaGSFnHYjGmKJMPGh+czEEZ4hheDbz9uViB9YBPSQxh7Df61Kd4F4Xa9XCM2RpYA4dr3+seudY/U8H2XScS7WiO8qfvu30voztMqgJ+/hzsMMm7OXnYI8wYXA3Jnb+U0MQHGl8dPYlI0C0+p3BYAfNMSxO1SM1DxJg5rSIV1n7WhSlzFSo/7Y9hOr14E+FrCqhYL6FF1EhNyX8AcU5hW/W9PovH+i1q0T1ECvw2pXRgiRu0U6BPCtLiRboBucxpNGNT5+jTbXKinbpwSq6FpCIRMZRJRKI0lTvH/JtIZerCiKVQCFS/qoSUEIuV/MMTUsgCDJcyIAps0V1ExWCCwQhyjfzVMaQZGu0CW6yIk1uZCK68K5RQ5aQSDRc5xtsCO3ObIONbwnLTVQUkaq2XdvPP2cbKFfZJk24C4WIhbxGGol2szBfXvbhNZPwhEigyk+4vzJTEKWZWrKd2+gDWwLPFqCyyodSCPhmVVX5uNcjI0BeizhbrzNVdrNi2eNecid7H0WR9YJuv/ewO1UNU/PbjaJlLkpf5UK9vbx40ovSShSKEF6xQXDA2NyUAr5Z2MoE8HCqqLjczBN5LVFyTDmv1ye8aNMCDSWZtgWy5Vov49waRHleZLdyrQ2aRZGtQePEZSNPoypTMoYyS2VSMvu0PVdZG0qUgKNY5lHMltJcQJFtlqt0C6VcyzQqQCoyClJxK2kgF1nRVuFkMA7QlNcarzGfTpQPfZ/UYZ6l22HS1iZTEd730ZZXnsfGrWFnR/x1I6+jFGn7/o+6i0fwwh7eQRNxr1fjYAm5FkUlbiHCiZUKWkTRzFS/HYRh39tvZHSolSvmLrWVyOsyjIVM26o38HwCDN89fffm4vdPn7x59bLteawyuJnR4Wa+/pJmRqaZy+8fP70z/r1BIJv4iqxQ3cJVnG1UVTdzB8zB4G+i/NXlu8P8OTvcyvmXtHJmWnl2+ert4WaC/sF2Bv0vaSfom4aevHry9LLZ0ieawD09ld2ZjH6B6yzd6K1LFMebIoq30K4KSVsyEIuFjKVQ8dYjEz2xs1WLcrXNzU7nchUVFVzcihImtz5sZxDCxGDiAfUh6EMPRFFkxYwqAbTpDUIIoEN9uGKSoAdRXG2iVL97SP4fIYRRb3jy+hlNqQsCQNPYEsTAU/USMfb7Zt+5jm4djQG4e+ce6WpxlMablPWWb9WNKCtUZiJhV06moKwilURFouuVU7VQ2NI50RMTCxJZohcTYlnEqWjso99ACIOzz8BGidzQLocN6KA3OJkEfuAHMyK3nYcTreuG42D8Em0BUbVp6K0+C9izqTwPxWyiTtqDE6sXOgPvdHDSVh1UEUGnlEuF/jLiGuHxen+czbwdPZd3b3Hsenl3O/M0se03rgeV6R18Mb2DXyR4hASrX5XYmr0BkstGYZPcAMntDfYJrudg4ANNQnVY11JfUqGo8iSY/br8rrswPJn0/eBIF4iNL3vD7pnndkVlxZrmv1W7gaN2zUAMTv5u5Btx+VuI318z/p8QP/zbif+MEJmV9B8lPQPsQ//AfMU+nLmUZ3EVrURSZOoqSxMmvZ60f3dJ+SyhhtlfxOrm0v8PnK7BQaHpj4PxyKVfxlm5y/F/GMOHxwl1GD76QunesYP+MTy3E/WzsvP1F3Vk1876+/aktt+0T1naLRc+Sd6PoVMZzyrmmwpPoJarKt06FpvxodO2jPAkYhFt0srueKCstqnoQht3u7QrQZuJtuAR5JlUlTdVZusIjfU78SAk7qGFk0AYkpdgAY+ggDEkvQF1mj0ZV+K2KjYJcySMirhdhG98PgQIg6/7frZYlKK6oPfO/b7Hg0kkv25uIY0PUW/z2RJkI8onD4OxPzvkvsQDCrRWbR8aSzbaqzSOMOkEfgflxZ/QQwefsICf2MCBWa/8a1G1h56va6GMDej7kP4O/KHfn/mTIU0jmMHsgdM2a3qnUTx3xf9xW85jRz+CfWTPj/7Wqb+ZR6hp5sd9Yg2RfX+oOzrwR/R35N+j8sA/p/d7+Fe3OPLPqHTknxPU0D/3zwj63B9S+ZDeR/7AH+512dW4TsfJ2cU09IlhvKoS+aagE+iHPrG+ZvmQqDas71O7+Hfk13w68wfUxzPd5zN/RFBnfuCP9qhsKCkkkwQbj7ECH+YQQjs4JWaeeV5v4AOeqKJB1Z5E/nymdQFMOpE/J5rrvx1+6Jgn6J3AMksTodDnVPGeSyr4tz+jr0QJOoTkkZ77fT+a+RP3bwcfOubpKLJ/30WG8P6cuFT/7fAD0TaDY8j+XCODGfTg1gwEOkHu+4BjASN6/Bofv6ZHlAg4o0cjoPdxe2vGBwZ+EGDEgT+BIT2ibME9fAyQY0GfSlHy4JweB3Xlr33EjeOKYQtnFgwfA/1Y48HG65axoG+bw8eRfhxZevDRdIbiInYFprkYaEUY7QjKyId52B66BUhD3jcTAQemi+R0z3w8sMLhqN9peAALOg6AW4AQhiUw98FURjDz3jHvnV2ATgPC0sMt0Seihxsy751dgM4OhKUnQjD+DzVn3zv8rp9rgLoAIQyezg6ezi6ezi6ejosH0TCyBbEdDN5gAD7gOAcjH3CQUUQmEJwDrsw+4NAHthgQOvjaB5Kscy62qO7pOvQB/0/CFdzHFoY+kHa+Z4rhXKNCmQ7uc7FFdabr4AcEgPuEaqSJJQE/s8WBoWqg+3HfGYCvDVX3NAkk3HBm+j3QneTiwDR8plsZOrzLA+QdWlALCGHR9+C63KzbVJJLLPIg709yOYOZZ+qg9VCaenkYZyqOqnbe9yEPPI8W7nbuMTBFsaSyglyoKlqim5qXbjyCy+yazqfGjs+aBpbgDH2SzOpAW3aLcNGfLOTM0+POBCMI2t2o5lU773unC+nDAunHP+1cngbeb89mwCyYKg8m3B2fW5vV5ol1dvMXn/tR4fmGSJamA3jUgmEPHBuEfnTTA/aJS5VvKt2T9aas0ANed9p6jyV+zxSdTTgusD0TzVqxfsJew4CNWNRU2CVkDTPkmPHm45BkOZ8YaPsKyB0YrUWF4U1YO8Ha8rok6HKVFRjPkfC3/BpCanRCy17PCATHBjFMZWEChrFuSDw5vm3AKnHTECloVxjFcq1NeCzNryfVpD+b+UBPgX0azLQNrmUA2iwm42Tmw3v92EExQRetPH3/VdiH3/0O5Fdhgn/ffxUmHvXxg8whK+SSIjbirMCTcyarPtVDesrN/DpkMq+pLeNX+hAmHdl574OU4UT67/0PZsqaf++DUMpJ+5qEcDjDLbeUk2s5O30f7IDm/TDHLzilwtytkw/0+4Ded6pdybBMRZHjbMwDX/US9CZfvXdKB1Tq1vNAf72Sfn713n8f9JQDMHOeaW6TikA+ePqLIeMd+k+NxLaVB+qkrU4DT0sdy8AGD31syCBVQ2l7p9rJaeAxYFZFqRErRzHgwRNOJS03KBrvMLxqSLWUuDF1MDDKDmZdI81pQVft/NrTJEdpaoXPKLJrv5ZJA4exVPJaqBqtlWgCRLmmSVz5YGvgrsWUS//9blfmUSnNtCl9+AEkhA/JY48zC//Q5vP9bilK3CfbSJwp1CC0s5uTe9ohSdyiSipEuUlJcy02qWYslKLq6gkt3zfGza986b/3QJ6+D8M+PAKcezAGGYYJvQX49t68DWYwNhKS5nAK6kThHxoaHtaO9OAU3kMH5yehDGAMA8Nco+gXaYQn222eXihMNIFxxKprr2N0P6uHiUJFf3Du4waknv0Bzf8JdtP2za/fToPmO4D//jTwoJ5bBnfwxbgJk/PeCTRuLSMzPbk8SOAhBPAIJm9OrEiaNQnGWGw0rtaqZqXik6Bf8d9UwckJPMfgTrFYCIqC9yFD0RIYIKiPh+vz6RI+iJzDaguhA/pgLtLspgsa3a/3r8dDgr22S+PnPeDBwLceDeT10YURP/LsDvXelIo43CEEs6PFsg2tSnpriwWJvB7wksmtanQ4flf14mZNFWi/N8Jtq8x8nP+7hZ6zpLv4INS0wikMun04gTYTegJt2UFycMZtcnzHt857z4NeTSBj9ODNCan0GrFXG4cplmqN6fTEa1qBWmvWM1mD6emzE0FH4hvuYavhOncFRO/SnQD7PuG8C+AROJqidkB3h3LBo3Y2I1W9M4CkLvbH1egysEMr++HihDh++v6kbQFPA+9UWk0kg1D2T2tjQQ7wvQFdfxuGcnAamIGeyL4vA18OfTnYt34bxm/T5tw5ndmZIX/jPPv/aVLlXzqXoAM7s8kyPR+Qudq9PUFPcN7dnuTdrZ93P57k3Y/1ilKgnynv3sIJkA8h6Pahg/t3bCIfdLdwCvmg+9Hjv9sTfOkNu/16eIstodgeQ/GRq95qFB8Rxe0Oio+E4uMxFLe6dY3iFlFsCUWtPSbFrV9s/eLj7O+kN+A+7tlJSOAuuuOusKg+7gqLcFqH3AkWjsH+l1IkuKui4wxngYb65NHRMweOIxNfYSxNeMdNrTZkd5QCwJ3d0L7xXDd3NDvDRA7Xkfaxj/TBAPufh/7I79cznBo80774M32AcEZwff9spxWUE3Q8s2ncZCWM6y3DFRW1lT26aoKyeJETl10z9T7dLvTol4GGCYk7Jwhd3wvICq90UNxfu3ZsGF4MUEOO/gO3iPXmP85UWRWbmAJsMFZw1+NcQlSiJ+FGARmQGDVZYtikDpe8kR8kXRKgwEl8630rUzF+Vp+cdhhnp8bZLa+XjhwdOmU1gnS3dSlfyb1DASxfR0slqw3dHuLzgcDPV3LGs4TOE3oWhsrwcCFfyZ3SXQm928lC77MnCwbTXY4VjmD69wamO50pHMb0ZwfT7JCiQyOLTnlG/n0fPfFf+/dpXpz59/W8ul/PowkdJvD0CgI6pdo9SMADAjpGqE8RvvbP9NQ805/PCPW579T09UNfY+0TVfec2czU4RmU4y/91Waqx5MUjs5SZ5L+UrAwsfzu8cFTxRWuErHOmhdVGh849DZiMw358sAINmmzEb/H+H6ftBq95yn6lAnRt0W2fk39bqN4xxwsiXGLQKpAfyPZj/08DSb92Qn+CWYzhqVA+DwN+I1jAPjM1u930Q9CaLCtGorP13mslkiLPoYvwhEfxH/tTfqaVgQZWL+jGYpl4EGVDRMcB4IzsCuy7qiSDxNZ4CW9dt83wQLLgdcJvJnhw+AAH+is1iw3fT6ApnVH86anezwwt2jKfI5kkfP5NbaRqcdb0+MBS9fKh77LoWmr4NtgfAGHYz+6/TMdg7rMjIuFMc9zkr0qG1CPm5jm5lbaXVENj6A6Omy2q9T1aQuxTFs+/AKJn3Y2K/sSh2dZ8b5nPBqi/wiHN/J8mNu3uedDbN9iYwDh4kiTMi6ysmzPh9CBaEig+OAZuBvUCyfRsF6ElX/DS+w+jd9GMcdM4B3QKBY+pFFZyXCwT65dMowDqLZMah/7hLBYe4KOdfeFj1FNEJS86e57sPNO9Mwcm4KQubrCHcNICZ8uJ4eToN/38f9aSn6ouUhgZuoxw7goMEVmC4g2tC1EBRTFWdlWcAKb3NyVwRMGHhLlO8V6Bk1bakNSpPz6m7bRoxAPEa/D60asNk7Im9rXTRte7NH+vUd2PznDupDoDbtSkWpfe1ab4B7vmg8ebsPw1oPb2mfX66FXDJ69vHz15s3Ty3fw7MXr509fPH357uLds1cvfXj59OmTt/Dk6ePvv/vu2cvvgBxzGCZdVZVY52hm5VlRoV2Gy8C/UsxPD+Jy2X1fAk6iPX0xbVkfL6EqRLUpVAmTRZFhPHec4WhExbfN18dR/AFzAMQfOMxbRPEKzb0qkgpdyhGksiRCjNcS46ujv7mTjvfhkMojSdNKj4R+f8aIHKdLIDpnej5cvnr9/OLlBVp+fV307ZtXL99RGAi/P764/APatvr17euLly+RpBCGuuhlhHHo/Z7B4J4cED0H5+Zqf3I2z7r4lJX6YTHyeau/P3P8xqTRAKk51GjsR7Hx15/XA7xFpcegfhzMrOZ7zdaORVX3kO4K2D7obWlKm1EzgXA8oCKlqDHLGXTgBuwUa1fwDXREXnrwiAdgjGUPQRfxGI3t8DU3Tbi9FFERr6wgGzlMhVpWK+ijIAZ4TU4Hx9WMWWbq3TZH1nAbJ8RARtfmMl/30UPvh6HY/UcENytiUV3PCANOJ2NYmHbDWsIeWeyOGmmehDT4Cw41rhsL/6HboC3hFAIPfgtp43SwQixE2wTPFav39vV98xjxWhK1esh8uH5fv7+f1ZAeVPIU0RzsDFni2OhXIbPqEUgY4yzy4Vqetq/fd66ld9Ju33TUybX0em11ogs9PCAZNxA18FxrRJaWmefhGYnpBuqq/9oMZ6muOfW3s3wX0zGeo9fPuPxQ4D0f0rkuwPGo7Yj2zrhQC57tiI3HILLNf3Bzxd93q2uZsH6XRjWj1Jy+7da3a8OjuvdtdcK67yH0dznsYD1E3h4VDmhjKN0K5EY5LKyu8E3SBTwMYQiPtFIZw05v07kBoDnQHDHjtd3dUM6LSKapoJ2kfj6U1+RX+EcmwXM5J8cOuG3pC2SXnEhDf4Ekq8ouwDs00uJCYugMJqPBe6QaggJokMoiIqNB318DuNwUGCmdbuGpWqayXMF3RZQIGNiq7aePL556FHS9ydGWEkkXuU15E+J0k4hyrJFJfodvHr96+3zQK6vEbKWbn11OCtXZlJ3l4JchXT4wIPLnuyLb5GN4zB9t8dvNeh0VW/uBeKTDo5lLuTHwfp0Ro5bfchqOutVX16K4luKGWn5dSFVRaiHqpIGRJURwU0gKPCi3ZSXWsCnpfngkS5Ew6dUqqkxOlUJEiblzTclTlhQjQljnW8hFlpsxl+s8khgrrq84A7yIlBsin2KoSYRxWhr5jb42TyQYpJr1ABcwbRnJiEWaTltIP5Ix59w9GJSlL49zX6hSowrjRE+oLCvKARPBEG5hAHRNfUlpkShETMv1ioYPb+NL9rbSlfdEqIpTy8y3GqcOlxlrMVhuCtEePPFfiMR/meE1TW8MF5jhA91wLk2aWVXJzDYZt+Zb7Ih04nCoGXSOBOcPmsxpT4LBcHR2PvPXolplSThtIa5py3tge2xcBZwgZFnI5CrOcinKNl79xmxT3f79wcmzl5e/x6gndB/NPF0bV0pxW+HF2KpoDzu/KbKb09/EWXoy9HzaUnbvnfmRildZEV4+ffnu6Rs/wnU0nLa2vDWztFxmKIibeua+UgJuoi2nTlrwxYqo5tAqQr0hCuA0S2RAWmapzVoUMo7SdNsFeKY4EUAq50VUmIHZZhuSryRjQZ5vYSmveVvE+bYglR8ETFuD0dm0hRZpZGIF6UMwGJ37cLOS8QpKzL+1kMIMB5fyBC8E5IUoBYX1kBRa4rUkaayYZYCDjqIKne4KcLu4daRBZSyAupWuw6o1HXuK21zEFfaAj2/WUZ5jT+aiuhFCsRKWCnd+epJxX4ympUhBwknzZ5fZKPqPNxVPJVltOOGCxPjIeAVrzMGkr/d2a2VSRR8sXzY5Np3RVKLUAlBmOHzr6AOvEKwQ51n2oTQJrTi/BcVKKT5zWhzSBDBtuavJtAU3UQmJuBZpltMCASYNlAMGhchTpAOTRogCNE4WgJK3JBX23AgDttuymk4WgMci5XY9z1IfhqNzO8ZRSQnDdA2zPvGAb/hbgxBM4mDiSu3gIDPW0RaVIDFW5/SYb6HYpJar0Rwza9yshAJxG8W4dDZQx5zoLY6UyiiMdVOKZFeI3jlThMfMRcLt8flQVN+/Nrw39XDKiQRf02wjLX0qwSZRKWel0KiqDPIiSzaxMONLbfEt7zW0p60lL/vowswW8P1bK6W6VV+jR2sA2uYjiZ2AC9YACp4mpKE9VAQVMVOgIG1QOSAvbEIV1C9JAlm1wlBawlyvRFkB3z99DO3vFev3Rmup8Ha5aRbWhkHEgxvBt09eYIYypdVXvEKuqaVUy3ppYcVxLYotzwMjmBlP31VWyI/oyUkhjbbZBjOakMxim5VQSalzdkRQimtBOiibp2KtL2SZRvT8MNNTJaKUBWU4LFdRkUMlc5z0tjsc8Uy8i/JcRIVphzSAiEpJWu2ZAiGRkTQPzEBhcLQQ6xI2CmdTumXNe/HkQkt3hIvspsK4a1nWI0Nss0wjNhlfnDbLDpl35tMBc45knW/pbeEiwXBQSsgTqQy+ia7Xo3+hMOI07Ypk81BbUv+arfDIjO1Nphx3nIYmjAE33IFyU+jQc9JK6MDDnuPUFSVn4EFHHwWKtzDBTYoToxRxphUw5Q41iUqiEl79gSrF25Re5xFbvGxVZAXFX2qZSpjSRXSdoZXltEN8nLbiLdlISrcHpABrdG8oDw6mtk0YI8abYg95aTaCiEKJ0bqSYs9xBRQJnK3XvdJOnq0oYCVoYeoGupRuHdJR9Y1MqhV0R4yUPpuT5Zubm66eOJzXCPPL4OEyGhWdSCUdbZx0NFSnXqJ2UGESQ4WiLpeqG2drY7PTtdBoKUzTT7KKU3Si8/s+BdP0z4dMZG3u4L9+tz84Y4DhPd0/AxGGD7HYZpBAA5wppas4jJdZQas6XhbFv7o70O8ORkGn3x32+4YtV5rgK8zOof/hlaSzexibja2Ru92CaYaHGoxusxwAS7LqqhTotCew+wMNdv+sAYbUabhJvzs4GzHY4F4DCjNh1VAEg1Aj4zXXxw5u4219P5TJ9enOqzdVdOaAfncqCMNaevUeuJlb0OZUc864GGOP8c88m5CDC+pUX5iZt93tB168mWcyaU+GJ5om+6BH9zSYGSv28at3jYSVDVJ5phuiKJ0KIzoJuoGuJQuKWYOOxq2La2K2absIwqzwi0EoC38Vag7Fq2i9EMUg7Pa/5pdILQfhqL9H2gHCcMZ7h1oKbU9NO0BpnyXi7g52cH86tql8oRNIoo4yp9ylPS/hr9Zt0Ob9w9utyvJSluNDvgNSrRHbq9oaZ5vL3SrzXWtrN5gtocH+LlqWY/hO6HvW32MOV7O7MKRgAz5MaLMywwdiGT7pLRM+aibgY1ZIoSoGzaUK8Zjs4t27i8vfP3357u2nBzoZmSXRtPeEM8zgvrrIbtCeOWhc4+JPEp4XooqMtYur9V+Q0r9AVCw3tBdNRE6rvL55ZDYgCfyFCf+LsUSe6fq6GBdWsw3ED8pgpj2/8cPrzSUKqjsEzT0nLYFUw5oT2UIfMjV3OKWwGCmZHBJsTEi7V9NYC6F3S9x9lANj76GJWwgIahje51dZDqlYVEiDD0Peu2RVla2d4pEF5QhR3L7X251zt5IF6BL7mqzbKImWg+FeRSOGFou5IxaBBmlKL5HeZI4xi0pjLEV6U0Syf95BEo2YGJz1xpds8e9PB1/3+zgR8Gn4bRfgW7zjgsfTlm7tRkHaO0J1loNpy5pkSu9Vab+iRUFn4cO+KGvp2j2J0yuz/b2RacrX44y7L+E7dvXsNLa6xvb9W1ge9Op1AV5mlWDhoryHrvm2O//JUjWM1wZfAidzscgKcbLjKiIEKE+I1u50Mkxh+KFrtmdm2lxw6+sMZxtuQQshODND6c5eHCJtsVK6PMcfiXN5mUWp9lBo59CWBQ2twEwJzNfI6RhtcuhClJUZnTKD6DqSKRvk3DjuYGLKMV3vKbXL1IAgJfXCWdugvjY8aTnoArzRxmy65c2jNWXZ1/ftkxeO/Wp8akj5IjOyu2fxUhN1Vgy3Cri7FgNfd8JynrQtpR8uOU+lvDYJKGEdFUtZZ2kmWZynGZ5z496WMmFyOjK+6anhrNsOfRUiUqVx/+g2jR9P+yCqbCloQtYEp1ugJOr19kdvo4zpVk+aOCswss707EKrbOvZIqJDrTWMW1ynIPONKjg47fyGavFZTzg6mxUpCSUn8tfLhN6e7C4NnU7HCBpaviH/caSbJdvSNG3xVpScYdMWpQ0nQXrCp7TjGkLj5QYhNA+ONq97nCntKTFTnWXWKlqfm+M1q9laXcM6braUnZ//NppjEXBJ1RPEsIOXebyZaEJLaAJyqb6XV4JWS/raQZuCwTwX8V/Y0/kXQ1EuMQ7gDQWwsCNyTf5uzq8OEdlbhPPfIbrFFhYVZWPHdl3EfY2RbRAIgbPpk4dUY89yqLKbqMCcLIwGm29Q9/1rpuypzo3fdl3RO9bRtPXzT//n55/+988//S/4+af/+fNP/+Pnn/7bzz/992mr9ig7w4RO6OHAO+KF9vGPb57PfOObZpPL6qgdVzUl7xwcQKr9sv60Fei//ELSEwxH/nB0ju9cej5tOS2h/vNhtyU40tS0RXqDDKRKlFV3D9by8rFcHuclLgg0LWE0osMrnbCfEu4Hw8E+Xrf28yz7ACn+5/mrV3/A/3z4LPzv8Sci4E+YnvcreIa5WTkhbHeXp5r4hhxgGkmOEbKqsMhuGiuf/iGJRdWJUrlUvEw6Vg7BYRJyzqHP7ItXMk2MUsZ5zyt1F+Ap1kHTGDmNHi4nySXnQzLrBBNvTzD05g0Tsw4xB8kM9196v764SdpDj8y59qg+tTDavv3u1etTPIU+ff7023dezbtXJIn13ufdqxcOnOMOsJi+/dOTXSTvbjIcoB3xBoP1i9GhBdJAyFK8h89w5tPx0b2oqihe0XnezqjSIQsFQdEZGIbSmZVOodnBQkG1+ACXTnHgMyLwBTJA04y1m937GMLINinxIAPby5TrRHWMCCMO9/qcfWM4cuVhk7eDM5fhETGizYEMDTePw3jcXcgjEtGs5I6fM3QvZJLopPwOkrvWfh6V1Vc79R5wjXqkzc9TuNtaq9+cdZU2unaVthrR2orciM+D4OPiYZwyqVhG6ZW2MPFSgBZAK9k1Fg7/5ApMhK5hFO3RpZ5rJpv1eoupYaKyFEXVluoKjZw24/KbmD3yfLenLW1imN2AtSr8Jngjf8JuC9Q93+2qRs8fpi22K461oauwG8ZIrjF+QrvS4N17cVu5hIwdoHqH+cggudJlNLBes6IBqTdfDFWzMrT95Ea+qin58UeQJfedKuFokEla72Y/7xPQemA61Sin05buivU4HeY0xudOakm05iTGMLbIEDUkIE4Gm2rTc2qgp7jq60TOKtEWbFgLODyCPoyBnW882XEv1dZzqSHtoSP0If5nJ0OHDpKmsCh96nOy5yedSJXM9Gn5nnv04Ed2u9WfZuw/JDfdUSCdyQ9Dhdq7Lt0ayuNpuKgOw/T4WP+07XapExzrlQbXg4q31HTShmb1Og4bXZC6HANIf/c7qN9QBPt4smcSuMgFn3/jD1LRqbf2m1AYgnAkiVdz+TnWN+QOCY1XIU7iunnP2/k1Di2dTkhovCJrEm3UacvzHvaNnvkuwyP06yiVCSRyKTFtCMmhH69w+uCLjzZMY+eBOHZ+6gIjElWYFUk7Xnkd/Evm6o6GcgLdj4/0yWSRZphMszf0/I767XC2gwSg6X2vTxXuJJTaP9/oALmZaTEvhGpb1/Mn6/Y1juDxnhbbcwDrRChmd9uwRhpuYdQdTf/uE9Q6x/y7V1Qp3COgrArvqK9WU1PWgSCOHWJIbLPnC70xJjTs4u3ls2fag1d7dc0aT/6wWpu6DiStWa0FZE9Ay40kF5dGwbnfOVU5ad0ffrAm1qdP2muDVpVV76yJjXmkbUHTzyO8Ob7p0z/N8JqP7UswGzsY0H8D/E/fByxk6wuNsMFwhs3buwGHhsJkCuN56oH5eIVsxNmxcxnHfscu4He+bRyv+HJfH1/GzZsF6KuKUpuTTIaTs3EnGPd1fimePvGqN/gP6f12gBl0AifUN0eD11aVyW0bkXkeBP3/kLP6ok12U57gl3qiHKbb9IvpZsqnLdALFnGZfomQhtyRK9sxrT02a6s/gv5gZK73a22mNuuHnDRLbdbfhOdDf9p6mSlzaArtc5QB75BvGX0SdgFvsltt1mR/2nn+Fm+z2ClLvkqpAH830bieCvypmvnW+pdMGJIOYpmqq5IckVb/VPpHFkuR4i9yNYp3l+X9f4gMx4iGS6iq2IYNDB50UNHTF0zUNKtz255M1Vwsb7IiGdvQIAwOE0uplFYF+BWzGtvgmLGZsjpWSC5VfTjAlp5QyR5SoZIa3ftMqiaENHyiYMysQO9ZmmY3HP9zs5KVYF2D+7NIt+3vupMJzggPGU1InkOH2vKOrVoJE9PBxx7VDbnGE3QiLqTCH+IpwWIoNTW806OgGd9tl0nTcWdTlWY3v9S7HH/wh37YDLtwpK9t9PLmOJ84+MybqjV61ne4KxWsedOF9oTL57U8DmphKOVWMlV5VFQ7w2bYxT99lEDbkt1zSebR8KYqLxa/1PGscPyKmthIk1Juvqx6Q1Atkm1KRwjj+uDM8Z5v9c9LRTEfiuV4naJQGA0pFLrVG8Fk030pLTcF+TAPCWYPXaf1cE2V/QmUHsAl4EolYvKtGhcuF6Jd1ShJZXWgMLsRBZ5cNUrXUbVqFCjMf0XxHgsKOYUOsXHFsVeFljVa7zkOl0ev/s0qlmYfEvrdTIzeM/MNEK3TlNtXlwQCcws2eV6TruM8rhhnnEZlST7sok2lppc+vtmK5rcv69VlXlzhtgE/6mWVGyyr4gr1S7uxyOICIlUS1vautaijUmtKvLqLP7mXhOFkBo8gXpkNZwMKL8qpBC+mzrzdTERIldqsr/hqF9LlI76+Xfto48bXTYgus35TTVnSqOMXskbxc3B6GOdp4HFFxPiNRdgJcAU0CHCd7U5bWHQA/2lALQyOtTDQLfR37BHkMINeSXVFA0h1gB5x5xAVVdj3QUOFfU/3Xn8Kbfd//NG9mIZ0EcTMJ1SexyaOxkM4xnds/zSw7Z8GB8cJg7OiuLpCaW3zYksC/jJai72MLnbLzZRODCBemOKqeDIy6dfmkbPJkirxwpAHjGVICxDfKG5SZsjCycoSZ4ij7j2XJZ6F0RIcTjDXzq58MXv12v9IgxpZ0d4PQ8SMLldaxJ5zu2j3ChkAmRD0C0t19WDmO2+DWX13qN6T3a1LE15bzSEctsW9s5J+HE+NRtfnag+OGE9TxbqHhh7CfWlommCoe1mlsvuFdzB3r6w267pyHOWlVn13qozwDM7152KJRbQY3qW+A88IhEq+CIEDzwia06jeedYmPl2pdbTtJF7Nan1rBw+zLOCUYVHZ/24kbK/NK1nyrOcJT+2iZy8RCyLid7+Dpm9Da5OH/Qd7iGgFZ9IbqPGZF8J45e1Xc5a+I5UdiMMoeFSP1HYWx8O1cTzmhYg+NBDUvfnxx8OU7mPiNeEIGdoQ0TTQCmtsGDY+4hXG8esDk4Lt2gTP4tl+AfppELkAg0lH2+eFpEBXDJOmgBqFgdtACPeG2+nCVRlHSgeySZXgT6QUVvklsghDWgBJDeLSQb9IzuqDvnbqz6Ql0Q3cBHMZiZ/tSrwHowfQAbJMZ542qmNahrr28S6dJrLgbu3I/UKqhCfGFbHJ6FCHGVU0by4HuqC5Ikya07amkn6ogLo5Y0opsTVbyCFPSUaIs1Lr96/KzRwXZE0S0kKk6GrY3jHKLTAjxYSz/Lt5VbyCBZpdlMnbFpWbmDcfaKffQGx9BgX+bCaGf9M90NuqJr5eALIcN/U7vei7F5zzQlyTDIdgWNJx3SKgxG21A3BKCQ10V3dhfwm+4XSh3ctbtFzqyWynuCHN24F/qpID0KZdF5pltYlfy+8h5PzJRa+BD+EmYW+iZvk/hJm+uIgZ1OLVl3/tNWg9buG0FaU30bY0fpoff3S/8eqEU7tmo355qhLPXfr36moXyIHqlgvewYp6I84VXQY3Wj5QUfsHXMOc9Y1l2O4Hyx7vED69zT9ISN2Hw11Xm/WBnptROtacreWMvVvtMLcO1HKGhv6Z+yrmunPJv8cbpXQXaCXnsjIXyhp70kNEfm582mb60A7EsOjwcBnHCCNqu5juUpt9IrpyzWVbVfPBFZoDaLRv5MBYHenKIVKyssKFh5F8VU9yLW3HFyX8UadOcBhpIfZwWtp+afU+jNK4cPDI7Mg/IydJJkp0U3AcJMZ2ZsUHRPloR/9r/39T/w9ohxySJ9q4C+zCRfuRPQsXUyjsImnorAlC+q6e990lte6l42g6rrqQsImzW/HhjmTehQLjhj3Q/EHlQ9K2txzUn+l8/bk29JrfjEJ/BGb7WLPPby6KR4i1zlgerT0k0OjwHXqz10IhcryakRxrALEwjPGUWAumbvcw8bXfz1WTB5s5Om53N6MOeTsOEA6uoebjr1k8F8oxHo+Ydo5lx1Vqq/ZII6fNodlpsfHV+d29/RNVJ+jDO3IuWh+zOtk4THixPolx8nIkX3KeeiDm5PCZat2062aVHD+zT4xzq+eXbgyMDYTUV3/HB+7+Onev69PVvdAWG7GLd2m2e7d39c3uDf5oN0ab80YW7PEJC3EJQeecY8DLHK9X6sh1g8KNjHPvydjo9+3eMXh9mAxCUZJn+q1FaW6fgLl+QDaBjRegpjns+dDxLyXbOzCQB8JdPX3y+0an64IJHveOzHnviM58+bjXPQE+w/g7HwZn5zMrwrRFuYxyakDxec9cxBFaMHgFPtuUQC5uvpFibt+LWxwqkwSBPOIMYHqLQ0ENlBncCLihYPylwDs+INVCKjwGS7OMIk1u8JpFvCnQalK4iVo1r2E1VMURWd/3OPr16VhIWsqve0spCPU7OpY5JaFRK3ioHiX6RgB1kvwsdG8k5stx5Jya4zEL5QlwXJt9J0zMoZGDVTA2jamlZ8+37sTAIdch1KGx9qY7noFOsONARZuCPXLo+vuqHl8OvTrommaKGqcZxEOzJmp86Ln4hb7Z8w5ny8462mrn08DzfvnMGDj/C+a11xxy3JKeH/j10O0ySFP7EImlsXwr/rqhn1LBHA/ICnK/60g5x/vYxdQd6RYlEpdgdhexYmJDPiFZFmtJh8XW5+j6o/U53JpvBBzgVX0i1GSLr8n2PL+WUhbbQ12tfddH50OTfW5HfXDI9A/7ik41PeQZcj2ymKrs8wPo1P7MMOmACVdO8auR0+aBj12oTce55sP+znHAL7J/l+uMx3PodHhOY2Cf7sDzJssdL7zvMnzPhWkPFC3zmSzivXsQcDfec+Vd7XawSwcjc3CNQsZfWYWErpA92+5gkIjj7KsJRYThPlbXuZVHFWqYA0COC0qJm1qlHwAdNEC1ZRHWepVCbxwc+M5QFEbKj15nMPsq7B9m8iMzq9BUuVqJKDG1fOrlfpgf/RvvTkYL2zhY2JvCthdspDKT/GYfcGPa6BFxO8Sj4kND7opFPSUf6GwTf/d/NmXFBSzEDRk0OkWEvoY3bb3D+4nDS8xsoxIIoiiC4OLiAgMrht0R/ukO7SWI7rT1wNQOdYBos2jPetKnBMfsLVpWzVHCP44fr2gnplP6PNcJe3gfb3JoEMUbJf+6Ee0Yfz599yys75FlyEGBeP5hDOrMIKe7bdpQ1aGMbH6a5umeL4cdGThj8RtzfyHraicAjo+uY8NQHBNZLvYitkziLgeH9b0dxFFHZ+ncXG0O/ZAKT7Cb0R+eg9X6AQ9T5oYaNWlCP2wLk+CcT1uha9SSDdFAwffCEXaDJOgR2UEzL1KiopYzPD+WeEXZGCd1tN9867y0dzF7DmrdqWkLU1voDtL+B0+qpBnI0uGIcVE3h49dLE5Zw/PTHKY92NpH4xQ6vhC31LiVm2N0CKv1GDtltZvTLTTeS6fMuiKdsv2Kta/xmOByEB5ma4o4Q7jjKSQVRBaRMGkLND3iP4W1YqRkozpoHddPiDfFeazRgrV5K+VSYco3vM2KqRNFvKGr2lyTrpUZYWO8j4hV1pPpREOttvlKNAOk/ovGhXGk5tFIMVs6pwLo6OBM/I2eMoul1Rm8h8ZRqjfL3W7X5tSiiE3OyiaVfrPpzh6Zq2X/puerWOeWdW5ZviocwtwvrvRTORaSCdosVckebl12ALf+soc7FepIDRPKiZzKcK9FV8Ixw67RsVnho89Ah66iM0UUKSUHYzRr2A0ZQaY7cZI6YyPH3pvZREewvBbSO9oDaUq7wR202TrfQ4m/z0A5BSo9YQ9r+zXsBqP8AmmGhHoR4DRfB2JgG/gP0ahjuz9DmY0Qgs7+rCAIGwYEHbPQ7kC4sT54ky5LhEZVTwsbuY15MCu61ppvPn7ERIRrvFb6DPMObjCli/XDcXahOMolZdDT0zqi9KFTde5ziIbSyWxpl8G/nYr/h/lGYVwWD8s5BWZJ/TvoGqUJpyqN3UjJRTV1OpfkBcZTj87OO0EX4E+rLafLiKpHdZBs69PMbxU4RsW3dH+7NW4t03nLb4lbTN2xV1gVm0QUl5jVtWyNJ7NPfgtz/bbGP7TSbFm2xrzZtJCvZfxBFLaYUwkiNDK6NW6x3mv5rUUWb8rWmLCJAttKZJXVNXW5eY03ZZWt5Udb9AkzUqdplJciubQf30VzJLP1NpUJYaVLD/PstjUjeGyg9U+L+8m9QdzyWzgt8AZza0z7S3p/u4qS7MYWoUv5JXtzTcuf/i+/7qx1G6YAAA==)), haven't taken the time to dig into either the cause nor the weird machine-dependent behaviour, but indeed I think that assert shouldn't have been there at the first place.

kintel replied:
Sounds good; we probably have better things to do than chasing flakes right now :)
I'll make a final read-through and try to break this, but I feel we're very close.

kintel reviewed (commented):
One more issue: The following example renders with default color in preview mode, but black in `render-colors` mode:
```
color() cube();
```

See also https://github.com/openscad/openscad/blob/master/tests/data/scad/3D/features/color-tests.scad

kintel commented on tests/CMakeLists.txt:
```diff
@@ -1035,6 +1047,9 @@ add_cmdline_test(stlexport             EXPERIMENTAL OPENSCAD SUFFIX stl FILES ${
 add_cmdline_test(3mfexport             EXPERIMENTAL OPENSCAD SUFFIX 3mf FILES ${EXPORT_3MF_TEST_FILES} ARGS --enable=predictible-output)
+add_cmdline_test(offexport             EXPERIMENTAL OPENSCAD SUFFIX off FILES ${EXPORT_OFFEXPORT_TEST_FILES} ARGS --enable=predictible-output)
+add_cmdline_test(color-offexport       EXPERIMENTAL OPENSCAD SUFFIX off FILES ${EXPORT_OFFEXPORT_TEST_FILES} ARGS --enable=predictible-output --enable=manifold --enable=render-colors)
+add_cmdline_test(colorpngtest          EXPERIMENTAL OPENSCAD SUFFIX png FILES ${COLOR_3D_TEST_FILES} ARGS --render --enable=render-colors --enable=manifold)
```
Since this new feature affects the `render()` module, we should run these tests also in preview mode. If you look at e.g. `color-tests2.scad`, it currently renders incorrectly in preview mode (the yellow color doesn't take precedence).

kintel commented:
Let me know if you think it's better to chase the stragglers in follow-up PRs; I can keep trying to break this for a while :)

kintel commented:
PS! I'm working on some relatively light test framework refactor to make it clearer how CGAL vs. Manifold is being tested and which expectations differ. "light weight" as in no real functional changes, but practically renames and moves a lot of tests around.
^ This could be a good argument to merge your PR and open issues against stragglers

ochafik commented:
@kintel Thanks for reviewing! I'd say let's indeed cover updates in follow up PRs, if you're happy to merge this for now

t-paul commented:
Awesome :tada:! That really should get a note on Mastodon too. Anyone wants to write some words? If not, I can give that a try :grin:. @ochafik any account I should tag there?

kintel commented:
Let me fix a couple of stragglers first; shouldn't take more than a day or two. This is also solid enough that we can make this enabled under the Manifold feature. I'll get that sorted too.

kintel commented:
@ochafik In case you have ideas how to address them: Issues I found related to manifold color support: #5216 #5217 #5218

Rami5743 commented:
If I understood this correctly this are great news! Thank you very very much.

I did not understand the phrase  (includes OFF color import / export, green subtracted faces).

ochafik commented:
> If I understood this correctly this are great news! Thank you very very much.

My pleasure!

> I did not understand the phrase (includes OFF color import / export, green subtracted faces).

This PR only adds support of 3D colors while importing to and from the OFF format. Other formats that could support colors, like 3MF, are not supported yet (and the STL format doesn't support colors anyway).

As for the green, faces that result from differences have always been colored green in preview mode, and that color is now there too in render mode (and exported to file, again if you export to OFF only for now):

<img width="690" alt="image" src="https://github.com/user-attachments/assets/f38cf725-1878-48cf-8758-13d7a937a4be">

Rami5743 commented:
Thank you very very much!
